Add gRPC performance section about reading data (#22001)
dotnet/AspNetCore.Docs: aspnetcore/grpc/performance.md

}  
```

Protobuf is a binary format so it can efficiently serialize large binary payloads without overhead. Compare this with text based formats like JSON which require [encoding bytes to base64](https://en.wikipedia.org/wiki/Base64) and adds 33% to the message size.

> When working with large `ByteString` payloads there are some best practices to avoid unnecessarily copies and allocations that are discussed below.

Reviewer: ```suggestion
When working with large `ByteString` payloads there are some best practices to avoid unnecessary copies and allocations that are discussed below.
```

`UnsafeByteOperations.UnsafeWrap` requires [Google.Protobuf](https://www.nuget.org/packages/Google.Protobuf/) version 3.15.0 or later.

## Read binary payloads

Data can be efficiently read from `ByteString` instances by using `ByteString.Memory` and `ByteString.Span` properties.

```csharp
var byteString = UnsafeByteOperations.UnsafeWrap(new byte[] { 0, 1, 2 });
var data = byteString.Span;

for (var i = 0; i < data.Length; i++)
{
    Console.WriteLine(data[i]);
}
```

These properties allow an app to read data directly from a `ByteString` without allocations or copies.

Most .NET APIs have `ReadOnlyMemory<byte>` and `byte[]` overloads, so `ByteString.Memory` can normally be used. However, there are circumstances where an app might need to get the data as a byte array. If a byte array is required then the `MemoryMarshal.TryGetArray<T>` method can be used to get an array from the `ByteString` without allocating a new copy of the data.

```csharp
var byteString = GetByteString();

ByteArrayContent content;
if (MemoryMarshal.TryGetArray<byte>(byteString.Memory, out var segment))
{
    content = new ByteArrayContent(segment.Array, segment.Offset, segment.Count);
}
else
{

>     content = new ByteArrayContent(byteString.ToByteArray());

Reviewer 2: Add a comment here, saying that ToByteArray will make a copy of the data because TryGetArray failed.

Author: The explaination below covers this. Comments in documentation code are discouraged because they aren't translated.

Reviewer 3: >  Comments in documentation code are discouraged because they aren't translated.

Code comments are fine and encouraged. Adding them to the text so they are translated is a major bonus. [We done a survey](https://github.com/MicrosoftDocs/feedback/issues/2515) and found out surprisingly, non-english speakers don't mind code comments, but they appreciate them in the translated portion of the doc. Having  them in both places is the best of everything. Often the MT (machine translation) is imperfect but they can figure it out from the english code comments.

Author: ok!

### Read binary payloads

Data can be efficiently read from `ByteString` instances by using `ByteString.Memory` and `ByteString.Span` properties.

```csharp
var byteString = UnsafeByteOperations.UnsafeWrap(new byte[] { 0, 1, 2 });
var data = byteString.Span;

for (var i = 0; i < data.Length; i++)
{
    Console.WriteLine(data[i]);
}
```

These properties allow code to read data directly from a `ByteString` without allocations or copies.

Most .NET APIs have `ReadOnlyMemory<byte>` and `byte[]` overloads, so `ByteString.Memory` is the recommended way to use the underlying data. However, there are circumstances where an app might need to get the data as a byte array. If a byte array is required then the <xref:System.Runtime.InteropServices.MemoryMarshal.TryGetArray%2A?displayProperty=nameWithType> method can be used to get an array from a `ByteString` without allocating a new copy of the data.

```csharp
var byteString = GetByteString();

ByteArrayContent content;
if (MemoryMarshal.TryGetArray(byteString.Memory, out var segment))
{
    content = new ByteArrayContent(segment.Array, segment.Offset, segment.Count);
}
else
{

>     content = new ByteArrayContent(byteString.ToByteArray());

Author: ```suggestion
    // TryGetArray didn't succeed. Fallback to creating a copy of the data with ToByteArray.
    content = new ByteArrayContent(byteString.ToByteArray());
```

`UnsafeByteOperations.UnsafeWrap` requires [Google.Protobuf](https://www.nuget.org/packages/Google.Protobuf/) version 3.15.0 or later.

### Read binary payloads

Data can be efficiently read from `ByteString` instances by using `ByteString.Memory` and `ByteString.Span` properties.

```csharp
var byteString = UnsafeByteOperations.UnsafeWrap(new byte[] { 0, 1, 2 });
var data = byteString.Span;

for (var i = 0; i < data.Length; i++)
{
    Console.WriteLine(data[i]);
}
```

These properties allow code to read data directly from a `ByteString` without allocations or copies.

Most .NET APIs have `ReadOnlyMemory<byte>` and `byte[]` overloads, so `ByteString.Memory` is the recommended way to use the underlying data. However, there are circumstances where an app might need to get the data as a byte array. If a byte array is required then the <xref:System.Runtime.InteropServices.MemoryMarshal.TryGetArray%2A?displayProperty=nameWithType> method can be used to get an array from a `ByteString` without allocating a new copy of the data.

```csharp
var byteString = GetByteString();

ByteArrayContent content;
if (MemoryMarshal.TryGetArray(byteString.Memory, out var segment))
{

>     content = new ByteArrayContent(segment.Array, segment.Offset, segment.Count);

Author: ```suggestion
    // Success. Use the ByteString's underlying array.
    content = new ByteArrayContent(segment.Array, segment.Offset, segment.Count);
```

}
else
{

>     // TryGetArray didn't succeed. Fallback to creating a copy of the data with ToByteArray.

Author: ```suggestion
    // TryGetArray didn't succeed. Fall back to creating a copy of the data with ToByteArray.
```
